70%: Introduce namespaces

rsinger opened the pull request:
This branch introduces namespaces to Tripod.

Feel free to argue about which classes should go in which namespaces and how they should be named.

The way I've divided them up currently (although this is totally open for discussion) is:

` Tripod `
* ~~` ITripod `~~ ` IDriver `
* ` ITripodStat `
* ` ChangeSet `
* ` ExtendedGraph `
* ` Labeller `
* ` Timer `
* ~~` ITripodSearchProvider ` (perhaps could just be ` iSearchProvider `, thoughts?)~~ ` ISearchProvider ` (formerly ` iTripodSearchProvider `)

` Tripod\Mongo ` (btw, it will take *forever* to get used to this inversion)
* ~~` Tripod `~~ ` Driver ` (formerly ` MongoTripod `)
* ~~` TripodBase `~~ ` DriverBase ` (formerly ` MongoTripodBase `)
* ` MongoGraph `
* ` Labeller ` (formerly ` MongoTripodLabeller `)
* ` ImpactedSubject ` (in current master, this is ` ModifiedSubject `, but that concept has been changed in this feature)
* ` Config ` (formerly ` MongoTripodConfig `)
* ` TriplesUtil `
* ` IndexUtils `
* ` NQuadSerializer ` (formerly ` MongoTripodNQuadSerializer `)
* ` MongoSearchProvider ` (this wasn't renamed because this is defining a Mongo-based search provider)
* ` Updates ` (formerly ` MongoTripodUpdates `)
* ` TransactionLog ` (formerly ` MongoTransactionLog ` - this might have been a mistake, as it's probably more like ` MongoSearchProvider `)
* ` SearchDocuments ` (formerly ` MongoTripodSearchDocuments `)

` Tripod\Mongo\Composites `
* ` IComposite ` (views, tables, and search are now defined as *composites*, this is an interface common among all)
* ` CompositeBase `
* ` Views ` (formerly ` MongoTripodViews `.  ~~Do the composites need their own part of the namespace? I feel that might be overkill~~)
* ` Tables ` (formerly ` MongoTripodTables `)
* ` SearchIndexer ` (formerly ` MongoTripodSearchIndexer `)

` Tripod\Mongo\Jobs `
* ` ApplyOperation ` (this is the background job to generate the composites - ~~the jobs could arguably have their own ` Tripod\Mongo\Jobs ` namespace. Discuss~~ moved into their own package)
* ` DiscoverImpactedSubjects ` (same as above)
* ` JobBase ` (same)

` Tripod\Exceptions `
* ` Exception ` (formerly ` TripodException `)
* ` CardinalityException ` (formerly ` TripodCardinalityException `)
* ` ConfigException ` (now it's own file, previously was bundled with ` MongoTripodConfig ` as  ` MongoTripodConfigException `)
* ` LabellerException ` (formerly ` TripodLabellerException `)
* ` SearchException ` (formerly ` TripodSearchException `)
* ` ViewException ` (formerly ` TripodViewException ` - is this MongoTripod-specific?)
* ` TimerException `

Where class names changed, if they had any associated getter methods, these would have changed, as well.  This was kind of a big undertaking, so there may be some that were missed, so feel free to point any out that you might run across.

Also, as I was going through the classes that were imported from legacy libraries (e.g. ` ExtendedGraph `), I cleaned things up a bit, adding docblocks, removing unused (or broken) methods, etc.

rsinger commented:
I do like that.  Part of me kind of thinks that ` ExtendedGraph ` could useful as its own library, but lets not go insane.

rsinger commented:
Yes, ` i ` was a typo.

rsinger commented:
@robotrobot I can easily buy ` Tripod\Mongo\Jobs `, but I think I'm on the fence on ` Composites ` - what's your conviction to it?

rsinger commented:
` Tripod ` -> ` Client ` ?

rsinger commented:
i.e. ` new \Tripod\Mongo\Client() `

rsinger commented:
(also, in what dialect is 'Tripod' a verb?  Is this a Staffordshire thing?)

rsinger commented:
Oh, I see, right.  I agree that the 'library' is Tripod, not the class.

rsinger commented:
I'm not sure I like ` IClient ` very much, though.  Could the interface name be ` ITripodClient ` (or Driver, I like that, too -- can't say I prefer one over the other)

rsinger commented:
Actually, the more I think of it, the more I like the argument behind 'mongo Driver for iTripod.

rsinger commented:
Ok, that's reasonable.  When spiking this out, I was concerned that it was just pushing the clutter to the namespaces (and that that might be harder for the uninitiated to navigate), but I don't have a strong opinion on that (especially now that I'm on the other end of the process)

malcyL commented on scripts/mongo/createTables.php:
```diff
@@ -82,10 +82,10 @@ function showUsage()
 function generateTables($id, $tableId, $storeName, $stat = null)
 {
-    $tableSpec = MongoTripodConfig::getInstance()->getTableSpecification($storeName, $tableId);
+    $tableSpec = TripodConfig::getInstance()->getTableSpecification($storeName, $tableId);
```
We seem to have \Tripod\Mongo\Config in other places, but TripodConfig here. Is this correctly named?

malcyL commented on src/classes/ExtendedGraph.class.php:
```diff
@@ -1643,7 +1783,7 @@ function get_label_for_uri($uri)
         if (!isset(self::$labelProperties))
         {
-            throw new TripodException('Please initialise ExtendedGraph::$labelProperties');
+            throw new \Tripod\Exceptions\Exception('Please initialise ExtendedGraph::$labelProperties');
```
Is \Tripod\Exceptions\Exception the right name here? If there is just one type of tripod exception, should it be up one level: \Tripod\Exception. If there are multiple exceptions for different reasons, then I guess there own package/directory/level makes sense - but then Exception would be a base exception?

rsinger replied:
This is the generic exception for Tripod.  I guess you could call it a base exception, but it's not like, say, an abstract base class.

It's intended to differentiate exceptions thrown within Tripod (that don't necessarily require their own fit-for-purpose Exception class) from general exceptions elsewhere.

malcyL commented on src/classes/Timer.class.php:
```diff
@@ -39,18 +46,18 @@ public function stop()
         if (is_null($this->start_time))
         {
-            throw new Exception('Timer: start method not called !');
+            throw new \Exception('Timer: start method not called !');
```
Should these be throwing something more specific? We have a \Tripod\Exception\Exception (but see earlier comment).

rsinger replied:
I'm not sure.  I can see the argument for creating a Tripod\Exceptions\TimerException, since it's something you might want to catch because it's not something to fail over.

rgubby replied:
I'd be tempted to go with something more specific (TimerException seems ok to me)

malcyL commented on src/exceptions/Exception.class.php:
```diff
@@ -0,0 +1,9 @@
+ * Class Exception
+ * @package Tripod
+ */
+class Exception extends \Exception {}
```
What is this adding over \Exception? The other exceptions in this namespace give something more specific in their name. Should this one?

rsinger replied:
See reasoning above.

malcyL commented on src/exceptions/LabellerException.class.php:
```diff
@@ -1,16 +1,27 @@
  * @codeCoverageIgnore
  */
-class TripodLabellerException extends TripodException {
+class LabellerException extends Exception {
```
Is this extending \Exception or \Tripod\Exceptions\Exception? I guess there is a right answer, but would naming make it clearer.

rsinger replied:
This is just how PHP namespaces work.  If the class name is 'bare' (i.e. no namespace) it is in the current namespace (in this case ` Tripod\Exceptions `).

malcyL commented on src/mongo/providers/MongoSearchProvider.class.php:
```diff
@@ -1,13 +1,26 @@
+ * Class MongoSearchProvider
+ * @package Tripod\Mongo
+ */
+class MongoSearchProvider implements \Tripod\ISearchProvider
```
\Tripod\Mongo\MongoSearchProvider => \Tripod\Mongo\SearchProvider ?

rsinger replied:
This is detailed in the PR comment.

malcyL commented:
This may be a step too far but...

* \Tripod\ITripod
* \Tripod\Mongo\Tripod

Before there was a namespace, perhaps being the main Tripod class, being called Tripod was clearer. Now these classes are in a namespace - the whole library is Tripod - what is the \Tripod\Mongo\Tripod class? Is there a better name for that interface/implementation which better reflects what it does rather than the library/project name?

rgubby commented on scripts/mongo/createViews.php:
```diff
@@ -93,8 +93,8 @@ function generateViews($id, $viewId, $storeName, $stat)
         print "Generating $viewId";
-        $tripod = new MongoTripod($viewSpec['from'], $storeName, array('stat'=>$stat));
-        $views = $tripod->getTripodViews();//new MongoTripodViews($tripod->storeName,$tripod->collection,$tripod->defaultContext);
+        $tripod = new \Tripod\Mongo\Tripod($viewSpec['from'], $storeName, array('stat'=>$stat));
```
Not sure about `\Tripod\Mongo\Tripod` - the one above which is ok `\Tripod\Mongo\Config` - but now we've removed a level of class, but we've added a dummy one in? Should it just be `\Tripod\Mongo`

rgubby commented on scripts/mongo/worker.inc.php:
```diff
@@ -6,4 +6,4 @@
 // this is so tripod itself uses the same logger
-MongoTripodBase::$logger = new \Monolog\Logger("TRIPOD-JOB",array(new \Monolog\Handler\StreamHandler('php://stderr', Psr\Log\LogLevel::DEBUG)));
\ No newline at end of file
+\Tripod\Mongo\TripodBase::$logger = new \Monolog\Logger("TRIPOD-JOB",array(new \Monolog\Handler\StreamHandler('php://stderr', Psr\Log\LogLevel::DEBUG)));
```
Just `\Tripod\Mongo\Base` ?

rgubby commented on src/ITripod.php:
```diff
@@ -1,4 +1,11 @@
 <?php
+
+namespace Tripod;
```
We've prefixed the namespace in another lib with `Talis` - do we want to do something similar here?
https://github.com/talis/persona-php-client/blob/master/src/Talis/Persona/Client/Login.php#L2

If we do it'll need an update to the composer.json file in order to work

rsinger replied:
I'm not terribly keen on this, but I'll defer to @robotrobot

rgubby commented on README.md:
````diff
@@ -30,9 +30,9 @@ Quickstart
 ```php
 require_once("tripod.inc.php");
 
````
Update the requirements section of the README to say > php 5.3

rsinger replied:
It already does in the branch this is merging into

rgubby commented on src/classes/Labeller.class.php:
```diff
@@ -346,18 +361,30 @@ function get_prefix($ns) {
 
 
         }
-        return $prefix;
```
does it need to return anything here?

rsinger replied:
There is no possible way to get to this line.

rgubby commented on src/classes/Timer.class.php:
```diff
@@ -39,18 +46,18 @@ public function stop()
     /**
      * Calculate difference between start and end time of event and return in milli-seconds.
+     * @throws \Exception
      * @return number time difference in milliseconds between stat and end time of event
```
stat -> start

rgubby commented on src/mongo/Config.class.php:
```diff
@@ -405,7 +407,7 @@ protected function validateTableSpecPart(array $spec, $depth=0)
                                  * check recursively through the predicate
                                  */
-                                $this->checkModifierFunctions($p, MongoTripodTables::$predicateModifiers);
+                                $this->checkModifierFunctions($p, Tables::$predicateModifiers);
```
Should this be `\Tripod\Mongo\Tables` ?

rgubby commented on src/mongo/Config.class.php:
```diff
@@ -445,10 +447,10 @@ protected function validateTableSpecPart(array $spec, $depth=0)
             }
 
-            $validComputingFieldFunctions = MongoTripodTables::$computedFieldFunctions;
+            $validComputingFieldFunctions = Tables::$computedFieldFunctions;
```
Tripod\Mongo\Tables ?

rgubby commented on src/mongo/Config.class.php:
```diff
@@ -538,34 +540,34 @@ protected function validateComputedFieldSpec($type, array $spec, array $availabl
 
         $this->validateSpecVariableReplacement($spec['if'][0], $availableFields);
-        if(isset($spec['if'][1]) && !in_array($spec['if'][1], MongoTripodTables::$conditionalOperators))
+        if(isset($spec['if'][1]) && !in_array($spec['if'][1], Tables::$conditionalOperators))
```
Tripod\Mongo\Tables ?

rgubby commented on src/mongo/Config.class.php:
```diff
@@ -581,7 +583,7 @@ protected function validateComputedConditionalSpec(array $spec, array $available
             elseif(is_array($spec['then']))
             {
-                $functions = array_intersect_key(array_keys($spec['then']), MongoTripodTables::$computedFieldFunctions);
+                $functions = array_intersect_key(array_keys($spec['then']), Tables::$computedFieldFunctions);
```
Tripod\Mongo\Tables ?

rgubby commented on src/mongo/Config.class.php:
```diff
@@ -603,7 +605,7 @@ protected function validateComputedConditionalSpec(array $spec, array $available
             elseif(is_array($spec['else']))
             {
-                $functions = array_intersect_key(array_keys($spec['else']), MongoTripodTables::$computedFieldFunctions);
+                $functions = array_intersect_key(array_keys($spec['else']), Tables::$computedFieldFunctions);
```
Tripod\Mongo\Tables ?

rgubby commented on src/mongo/ImpactedSubject.class.php:
```diff
@@ -57,7 +58,7 @@ public function __construct(Array $resourceId, $operation, $storeName, $podName,
         else
         {
-            throw new Exception("Invalid operation: $operation");
+            throw new \Exception("Invalid operation: $operation");
```
\Tripod \Exceptions\Exception instead of an \Exception?

rgubby commented on src/mongo/Config.class.php:
```diff
@@ -1803,17 +1805,17 @@ public function getCollectionForTable($storeName, $tableId, $readPreference = Mo
-     * @throws MongoTripodConfigException
-     * @return MongoCollection[]
+     * @throws \Tripod\Exceptions\ConfigException
+     * @return \MongoCollection[]
```
remove the square brackets

rsinger replied:
Nope.  It returns an array of MongoCollections.

rgubby commented on src/mongo/Config.class.php:
```diff
@@ -1851,10 +1853,10 @@ public function getCollectionsForTables($storeName, array $tables = array(), $re
-     * @throws MongoTripodConfigException
-     * @return MongoCollection[]
+     * @throws \Tripod\Exceptions\ConfigException
+     * @return \MongoCollection[]
```
remove the square brackets

rgubby commented on src/mongo/Config.class.php:
```diff
@@ -1892,10 +1894,10 @@ public function getCollectionsForViews($storeName, array $views = array(), $read
-     * @throws MongoTripodConfigException
-     * @return MongoCollection[]
+     * @throws \Tripod\Exceptions\ConfigException
+     * @return \MongoCollection[]
```
remove the square brackets

rgubby commented on src/mongo/Tripod.class.php:
```diff
@@ -596,23 +625,23 @@ public function getComposite($operation)
     /**
      * For mocking
-     * @return MongoTripodLabeller
+     * @return Labeller
```
Not sure what should be documented here: Labeller or \Tripod\Mongo\Labeller ?

rsinger replied:
PHPDoc seems smart enough to know that bare classes are in the same namespace.

rgubby commented on src/mongo/delegates/SearchDocuments.class.php:
```diff
@@ -1,19 +1,25 @@
      * @internal param \MongoDB $db
      */
-    function __construct($storeName, MongoCollection $collection, $defaultContext, $stat=null)
+    function __construct($storeName, \MongoCollection $collection, $defaultContext, $stat=null)
```
make this public?

rgubby commented on src/mongo/delegates/Tables.class.php:
```diff
@@ -56,23 +62,23 @@ class MongoTripodTables extends CompositeBase
      * todo: MongoCollection -> podName
      */
-    function __construct($storeName,MongoCollection $collection,$defaultContext,$stat=null)
+    function __construct($storeName,\MongoCollection $collection,$defaultContext,$stat=null)
```
make this public

rgubby commented on src/mongo/delegates/Updates.class.php:
```diff
@@ -1,16 +1,22 @@
-     * $var MongoTransactionLog
+     * $var TransactionLog
      */
     private $transaction_log = null;
```
looks a bit odd having this as snake_case rather than camelCased

rgubby commented on test/unit/mongo/MongoTripodViewsTest.php:
```diff
@@ -251,7 +253,7 @@ public function testGenerateViewWithCountAggregate()
 
-        $actualView = MongoTripodConfig::getInstance()->getCollectionForView('tripod_php_testing', 'v_counts')->findOne(array('_id'=>array("r"=>'http://talisaspire.com/works/4d101f63c10a6',"c"=>"http://talisaspire.com/","type"=>'v_counts')));
+        $actualView = \Tripod\Mongo\Config::getInstance()->getCollectionForView('tripod_php_testing', 'v_counts')->findOne(array('_id'=>array("r"=>'http://talisaspire.com/works/4d101f63c10a6',"c"=>"http://talisaspire.com/","type"=>'v_counts')));
 //        var_dump($actualView); die;
```
remove the var_dumps

rsinger commented:
From the feedback, I have:
* Renamed ` Tripod ` to ` Driver `, ` TripodBase ` to ` DriverBase `, and ` ITripod ` to ` IDriver `
* Added ` \Tripod\Mongo\Jobs ` and ` \Tripod\Mongo\Composites ` and moved the related classes into them
* Added ` TimerException ` to ` \Tripod\Exceptions `

Any other comments before I merge into the feature branch?

rsinger commented:
Should we not?  It's the interface for a Tripod Driver, right?

rsinger commented:
Agreed.  I am certainly not married to the change: I just thought it seemed more consistent.  I could be just as happy (well, ambivalent, really) with either name.
